File: rewrite-experiment/hypothesis-b/src/rules/impl/MarketHoursRule.ts
import { BaseRule } from '../Rule';
import { RuleContext, setAttribute, addWarning } from '../../domain/RuleContext';
import type { RuleResult } from '../Rule';

// RUL-012: Business — market hours check (9:30-16:00, weekdays)
// Priority 80, Category: Business, Behavior: PASS_WITH_QUEUE
// BUG-003: Uses server local time, not Eastern Time (preserved)
// BUG-020: Sets queued=true attribute but no downstream code actually defers execution
export class MarketHoursRule extends BaseRule {
  name = 'MarketHoursRule';
  priority = 80;
  category = 'Business';
  failOpen = true;

  private readonly MARKET_OPEN_HOUR = 9;
  private readonly MARKET_OPEN_MINUTE = 30;
  private readonly MARKET_CLOSE_HOUR = 16;
  private readonly MARKET_CLOSE_MINUTE = 0;

  private timeProvider: (() => Date) | null = null;

  setTimeProvider(fn: () => Date): void {
    this.timeProvider = fn;
  }

  evaluate(ctx: RuleContext): RuleResult {
    // BUG-003: Uses server local time instead of Eastern Time
    const now = this.timeProvider ? this.timeProvider() : new Date();
    const day = now.getDay();
    const hour = now.getHours();
    const minute = now.getMinutes();

    const isWeekend = day === 0 || day === 6;
    const currentMinutes = hour * 60 + minute;
    const openMinutes = this.MARKET_OPEN_HOUR * 60 + this.MARKET_OPEN_MINUTE;
    const closeMinutes = this.MARKET_CLOSE_HOUR * 60 + this.MARKET_CLOSE_MINUTE;

    const isOutsideHours = currentMinutes < openMinutes || currentMinutes >= closeMinutes;

    if (isWeekend || isOutsideHours) {
      // BUG-020: Sets queued=true but nothing downstream actually queues
      setAttribute(ctx, 'queued', true);
      addWarning(ctx, `Order placed outside market hours (${hour}:${String(minute).padStart(2, '0')}, day=${day})`);
    } else {
      setAttribute(ctx, 'queued', false);
    }

    return 'PASS'; // PASS_WITH_QUEUE: never rejects
  }
}

File: common-lib/src/com/bigcorp/common/rules/impl/MarketHoursRule.java
package com.bigcorp.common.rules.impl;

import com.bigcorp.common.rules.Rule;
import com.bigcorp.common.rules.RuleContext;
import java.util.Calendar;

/**
 * Checks if the order is placed during market hours.
 * 
 * Market hours: 9:30 AM - 4:00 PM Eastern.
 * Except we don't actually convert to Eastern time, we just use
 * the server's local time and hope for the best.
 * 
 * This rule was disabled for 6 months in 2001 because the server 
 * clock was wrong and nobody noticed until a client complained they
 * couldn't trade at 2 PM.
 * 
 * @author Bob
 * @since 1.0
 */
public class MarketHoursRule implements Rule {

    // market opens at 9:30 AM (server local time... not Eastern)
    private static final int MARKET_OPEN_HOUR = 9;
    private static final int MARKET_OPEN_MINUTE = 30;
    // market closes at 4:00 PM
    private static final int MARKET_CLOSE_HOUR = 16;
    private static final int MARKET_CLOSE_MINUTE = 0;

    // set to false to disable this rule without restarting the app
    // (added after "the incident" in 2001)
    private boolean active = true;

    public String getName() {
        return "MarketHours";
    }

    public int getPriority() {
        return 80;
    }

    public boolean isActive() {
        return active;
    }

    public void setActive(boolean active) {
        this.active = active;
    }

    public boolean evaluate(RuleContext context) {
        Calendar cal = Calendar.getInstance();
        int dayOfWeek = cal.get(Calendar.DAY_OF_WEEK);

        // check weekend
        if (dayOfWeek == Calendar.SATURDAY || dayOfWeek == Calendar.SUNDAY) {
            context.addWarning("Order placed on weekend - will be queued for Monday");
            // don't reject, just warn
            // (originally this rejected, but sales complained)
            context.setAttribute("queued", Boolean.TRUE);
            return true;
        }

        int hour = cal.get(Calendar.HOUR_OF_DAY);
        int minute = cal.get(Calendar.MINUTE);

        int currentMinutes = hour * 60 + minute;
        int openMinutes = MARKET_OPEN_HOUR * 60 + MARKET_OPEN_MINUTE;
        int closeMinutes = MARKET_CLOSE_HOUR * 60 + MARKET_CLOSE_MINUTE;

        if (currentMinutes < openMinutes || currentMinutes >= closeMinutes) {
            context.addWarning("Order placed outside market hours (" + hour + ":" + minute + ") - will be queued");
            context.setAttribute("queued", Boolean.TRUE);
            return true; // don't reject, queue instead
        }

        return true;
    }

    public void execute(RuleContext context) {
        // nothing to execute
    }
}
